### PR TITLE
REF/ENH: Constructors for DatetimeArray/TimedeltaArray

### PR DESCRIPTION
**Pull request by jbrockmendel**
Big push on the constructors for DatetimeArray/TimedeltaArray, some progress de-duplicating code from their Index counterparts.

As discussed elsewhere, adds `dtype` to `PeriodArray.__init__`

@TomAugspurger can you confirm that the `_from_sequence`s implemented here handle the right cases?  It wasn't obvious if object-dtyped array/indexes were supposed to be handled there, but combining those cases was too clean to overlook.

Small fix in DatetimeArray comparison methods, just enough to make the tests work.  A separate PR forthcoming PR will do a more thorough fix/improvements of those.

One non-obvious point on which input would be especially welcome is how/when to use the `copy` kwarg in such a way as to copy at-most-once.  (related: deep_copy_if_needed and #21907)

#23491 found during this process, will be addressed in a follow-up.

### CODE REVIEW

File: pandas/core/arrays/datetimes.py [hunk 1]
```diff
@@ -170,6 +171,8 @@ class DatetimeArrayMixin(dtl.DatetimeLikeArrayMixin):
     # Constructors
 
     _attributes = ["freq", "tz"]
+    _freq = None
+    _tz = None
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimes.py]: These will be needed later, are orthogonal to the rest of this PR.

File: pandas/core/indexes/datetimes.py [hunk 2]
```diff
@@ -199,10 +198,11 @@ def _join_i8_wrapper(joinf, **kwargs):
 
     _engine_type = libindex.DatetimeEngine
 
-    tz = None
+    _tz = None
```
Comment (jbrockmendel) [inline pandas/core/indexes/datetimes.py]: A `tz` property is defined below

File: pandas/core/indexes/datetimes.py [hunk 3]
```diff
@@ -199,10 +198,11 @@ def _join_i8_wrapper(joinf, **kwargs):
...
     _freq = None
     _comparables = ['name', 'freqstr', 'tz']
     _attributes = ['name', 'freq', 'tz']
+    timetuple = None
```
Comment (jbrockmendel) [inline pandas/core/indexes/datetimes.py]: This is another we'll-need-this-later (when moving from inheritance to composition)

File: pandas/core/indexes/datetimes.py [hunk 4]
```diff
@@ -320,19 +321,9 @@ def __new__(cls, data=None,
...
-
+        dtl.maybe_define_freq(freq_infer, subarr)
         return subarr._deepcopy_if_needed(ref_to_data, copy)
 
-    def _convert_for_op(self, value):
-        """ Convert value to be insertable to ndarray """
-        if self._has_same_tz(value):
-            return _to_m8(value)
-        raise ValueError('Passed item and index have different timezone')
-
```
Comment (jbrockmendel) [inline pandas/core/indexes/datetimes.py]: Just moving this out of the Constructors section for readability

File: pandas/tests/arrays/test_datetimes.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        tz = tz_naive_fixture
+
+        # arbitrary DatetimeIndex; this should work for any DatetimeIndex
+        #  with non-None freq
+        dti = pd.date_range('2016-01-1', freq='MS', periods=9, tz=tz)
+        expected = DatetimeArray(dti)
+        assert expected.freq == dti.freq
+        assert expected.tz == dti.tz
+
+        # broken until ABCDatetimeArray and isna is fixed
```
Comment (jbrockmendel) [inline pandas/tests/arrays/test_datetimes.py]: This will be done in the forthcoming PR mentioned in the OP.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimes.py [hunk 6]
```diff
@@ -207,26 +215,46 @@ def __new__(cls, values, freq=None, tz=None, dtype=None):
         # if dtype has an embedded tz, capture it
         tz = dtl.validate_tz_from_dtype(dtype, tz)
 
-        if isinstance(values, DatetimeArrayMixin):
+        if lib.is_scalar(values):
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: the array class doesn't have scalar_data_error; this is a copy/paste of how DatetimeIndex currently does it.

File: pandas/core/arrays/datetimes.py [hunk 7]
```diff
@@ -207,26 +215,46 @@ def __new__(cls, values, freq=None, tz=None, dtype=None):
...
-                result.freq = to_offset(inferred)
+        dtl.maybe_define_freq(freq_infer, result)
 
         # NB: Among other things not yet ported from the DatetimeIndex
         # constructor, this does not call _deepcopy_if_needed
         return result
 
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=None, copy=False):
+        # list, tuple, or object-dtype ndarray/Index
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: You're right we could make do without it.  I like doing this explicitly because to_datetime is already overloaded and circular.

File: pandas/core/arrays/timedeltas.py [hunk 8]
```diff
@@ -127,22 +124,44 @@ def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
...
+            values = values.astype(_TD_DTYPE)
 
+        result = cls._simple_new(values, freq=freq)
+        dtl.maybe_define_freq(freq_infer, result)
         return result
 
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=_TD_DTYPE, copy=False):
+        # list, tuple, or object-dtype ndarray/Index
+        values = np.array(scalars, dtype=np.object_, copy=copy)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: This doesn't call to_timedelta, so this does require that we pass an object array.

File: pandas/core/arrays/timedeltas.py [hunk 9]
```diff
@@ -413,3 +432,21 @@ def _generate_regular_range(start, end, periods, offset):
...
+
+    Parameters
+    ----------
+    dtype : dtype
+
+    Raises
+    ------
+    ValueError
+    """
+    if dtype != _TD_DTYPE:
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: When called from _simple_new this is internal so AssertionError would make sense, but it is also called from `__new__` so is in principle user-facing.

Either way I need to add tests for this.

**Comment by jbrockmendel** [comment]
Good comments, thanks.  Also some linting mistakes I need to fix.  I'll make another pass and comment when this is ready to be looked at.

File: pandas/core/arrays/datetimes.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/datetimes.py]: the exception is wrong

pls centralize this

File: pandas/core/arrays/datetimes.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/datetimes.py]: this is horribly inefficient and unnecessary

File: pandas/core/arrays/timedeltas.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/timedeltas.py]: then it should
let’s not reinvent the wheel

File: pandas/core/arrays/timedeltas.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/timedeltas.py]: well this should never happen all conversions should be before this

so it should assert

File: pandas/core/arrays/datetimes.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: sure

File: pandas/core/arrays/datetimes.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: If we don't do it here, to_datetime is going to do this.  It may be unnecessary, but it is not horribly inefficient.  What _is_ a code smell is the circularity involved in calling to_datetime.

File: pandas/core/arrays/timedeltas.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: No, it shouldn't.  `to_timedelta` will just end up calling `array_to_timedelta64` like this does, but only _after_ doing a bunch of unecessary dtype checks.

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: Besides this is what `TimedeltaIndex.__new__` currently calls

File: pandas/core/arrays/timedeltas.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/timedeltas.py]: no my point is there should be and i think currently there is already conversion

if it’s wrong at this point it’s not a user error but an incorrect path taken

File: pandas/core/arrays/datetimes.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/datetimes.py]: then just call array_to_datetime and don’t force the conversion to array

File: pandas/core/arrays/timedeltas.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: My point is that this check function is called two times, one of which is the very first thing in `TimedeltaArray.__new__`.

File: pandas/core/arrays/datetimes.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: array_to_datetime expects an object ndarray.  I'll see about calling array_to_datetime directly instead of to_datetime.

**Comment by jbrockmendel** [comment]
Pushed commits fixing tests, also implemented `maybe_validate_freq`.  I think this can be handled more cleverly in some cases, will look into this.  There was also an Issue about removing the `verify_integrity` argument from some of the constructors.  Now would be decent time to do so.

File: pandas/core/indexes/datetimes.py [hunk 10]
```diff
@@ -298,41 +298,22 @@ def __new__(cls, data=None,
...
                 data = data.astype(np.int64, copy=False)
-            subarr = data.view(_NS_DTYPE)
 
+        subarr = data.view(_NS_DTYPE)
         assert isinstance(subarr, np.ndarray), type(subarr)
         assert subarr.dtype == 'M8[ns]', subarr.dtype
 
         subarr = cls._simple_new(subarr, name=name, freq=freq, tz=tz)
-        if dtype is not None:
-            if not is_dtype_equal(subarr.dtype, dtype):
```
Comment (jbrockmendel) [inline pandas/core/indexes/datetimes.py]: This check is made unnecessary by the `tz = dtl.validate_tz_from_dtype(dtype, tz)` above (implemented a couple PRs ago)

**Comment by TomAugspurger** [comment]
> @TomAugspurger can you confirm that the _from_sequences implemented here handle the right cases? It wasn't obvious if object-dtyped array/indexes were supposed to be handled there, but combining those cases was too clean to overlook.

It should handle any sequence where the scalar types are instances of `ExtensionArray.dtype.type` or NA. So yes, object-dtype arrays should be handled.

**Review by TomAugspurger — COMMENTED** [review]
@jbrockmendel could you summarize where this is in the overall TDA/DTA refactor, and how it gets us closer to the goal (primarily disentangling inheritance -> composition? Anything else?)

File: pandas/core/indexes/timedeltas.py [hunk 11]
```diff
@@ -157,40 +156,26 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
...
-        if data.dtype == np.object_:
-            data = array_to_timedelta64(data)
-        if data.dtype != _TD_DTYPE:
-            if is_timedelta64_dtype(data):
-                # non-nano unit
-                # TODO: watch out for overflows
-                data = data.astype(_TD_DTYPE)
-            else:
-                data = ensure_int64(data).view(_TD_DTYPE)
+        arr = TimedeltaArrayMixin(data, freq=freq)
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/timedeltas.py]: I don't think we should replace the handling of object dtype with `TimedeltaArray` constructor (this should not be able to handle object dtype eventually), but if needed that is fine to leave for a later PR that actually does the split / cleans up the TimedeltaArray constructor

File: pandas/tests/arrays/test_datetimes.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+    def test_scalar_raises_type_error(self):
+        # GH#23493
+        with pytest.raises(TypeError):
+            DatetimeArray(2)
+
+        with pytest.raises(TypeError):
+            pd.DatetimeIndex(pd.Timestamp.now())
+
+    def test_from_sequence_requires_1dim(self):
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/test_datetimes.py]: This can be a test in the EA base tests. 

Although, thinking about it now, I am not sure we should require implementors to handle this case, as the method should never be called with 2D data to begin with. 

What is the reason you added handling of this to that method?

File: pandas/tests/arrays/test_datetimes.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+            pytest.xfail(reason="Casting DatetimeIndex to object-dtype raises "
+                                "for pd.Index and is incorrect for np.array; "
+                                "GH#24391")
+
+        # arbitrary DatetimeIndex; this should work for any DatetimeIndex
+        #  with non-None freq
+        dti = pd.date_range('2016-01-1', freq='MS', periods=9, tz=tz)
+
+        # Fails because np.array(dti, dtype=object) incorrectly returns Longs
+        result = DatetimeArray(np.array(dti, dtype=object), freq='infer')
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/test_datetimes.py]: I think I mentioned above, but I don't think `DatetimeArray` constructor should handle this case? 

That is what we discussed in https://github.com/pandas-dev/pandas/issues/23212 (although, @jbrockmendel , you didn't really react there)

File: pandas/core/arrays/datetimes.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: Sure.  Though I'd like to collect things like this in a Get It Out Of The Way PR before long

File: pandas/core/indexes/datetimes.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: I am fairly certain that the version currently here is a mistake, since it is overriden by the property definition of `tz` below.  _freq and _tz are set in _simple_new, so I think the idea of also defining them on the class is to make introspecting what attributes exist easy.

File: pandas/core/indexes/datetimes.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: The stdlib datetime.datetime comparison methods check if this attribute exists and if so return `NotImplemented`, otherwise raise TypeError for non-datetime objects

File: pandas/core/indexes/datetimes.py (hunk 10 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: Nope, this isn't quite right.  Will revert/fix.

File: pandas/core/arrays/datetimelike.py [hunk 14]
```diff
@@ -888,6 +907,47 @@ def maybe_infer_freq(freq):
     return freq, freq_infer
 
 
+def maybe_define_freq(freq_infer, result):
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: Yah I'm not wild about mutating outside of `__init__/__new__`, will take another look at de-duplicating this code without this code smell.

File: pandas/core/arrays/timedeltas.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: Reasonable.  But hey, its a nice docstring.

File: pandas/core/arrays/datetimes.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: > So is the root problem (referenced in your "circularity" comment, and down below in TimedeltaIndex.__new__) that to_datetime / to_timedelta returns an Index instead of an EA?

It's not the fact that it's an Index so much as that it is a circular dependency.  I think I can resolve this in an upcoming commit.

> Looking through to_datetime and _convert_listlike_datetimes, I don't see a conversion to ndarray[object].

_convert_listlike_datetimes calls `ensure_object`.

> Sorry, to_datetime has in intermediate datetime64[ns] -> object -> datetime64[ns] conversion? That seems unnecessary.

Not sure what you're referring to.  As implemented _from_sequence is specifically for list, tuple, or object-dtype NDArray/Index.  datetime64-dtype goes through a different path.

File: pandas/core/arrays/datetimelike.py [hunk 15]
```diff
@@ -801,6 +801,25 @@ def _evaluate_compare(self, other, op):
...
 
+def scalar_data_error(scalar, cls):
+    """
+    Produce the error message to issue when raising a TypeError if a scalar
+    is passed to an array constructor.
+
+    Parameters
+    ----------
+    scalar : object
+    cls : class
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: sure

File: pandas/core/arrays/datetimes.py [hunk 16]
```diff
@@ -193,11 +196,16 @@ def _simple_new(cls, values, freq=None, tz=None, **kwargs):
         result._tz = timezones.tz_standardize(tz)
         return result
 
-    def __new__(cls, values, freq=None, tz=None, dtype=None):
+    def __new__(cls, values, freq=None, tz=None, dtype=None, copy=False):
+        if isinstance(values, (list, tuple)) or is_object_dtype(values):
+            values = cls._from_sequence(values, copy=copy)
+            # TODO: Can we set copy=False here to avoid re-coping?
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: > it is not (yet) possible to simply remove this case? 

Not if we want to share the extant arithmetic tests (which we do)

> (eventually we should not call the DatetimeArray constructor with an array-like of scalars)

I don't share this opinion, would prefer to delay this discussion until it is absolutely necessary.

File: pandas/core/arrays/datetimes.py [hunk 17]
```diff
@@ -207,26 +215,46 @@ def __new__(cls, values, freq=None, tz=None, dtype=None):
...
-        if isinstance(values, DatetimeArrayMixin):
+        if lib.is_scalar(values):
+            raise TypeError(dtl.scalar_data_error(values, cls))
+        elif isinstance(values, ABCSeries):
+            # extract nanosecond unix timestamps
+            if tz is None:
+                # TODO: Try to do this in just one place
+                tz = values.dt.tz
+            values = np.array(values.view('i8'))
+        elif isinstance(values, DatetimeArrayMixin):
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: > DatetimeArrayMixin -> cls?

No.  For the moment we are still using inheritance, so this would mess up for DatetimeIndex == DatetimeArray.  When we change to composition this check will have to become `isinstance(values, (DatetimeArray, ABCDatetimeIndex))`

File: pandas/core/arrays/timedeltas.py [hunk 18]
```diff
@@ -180,6 +203,23 @@ def _generate_range(cls, start, end, periods, freq, closed=None):
 
         return cls._simple_new(index, freq=freq)
 
+    # ----------------------------------------------------------------
+    # Array-Like Methods
+    # NB: these are appreciably less efficient than the TimedeltaIndex versions
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: > If we do so, we should add them to all our EAs, or actually even to the EA interface

I'm not necessarily opposed to this, but this isn't obvious to me.

> Because of (lack of) caching? This comment makes it seems like it's slower in general, when (if it's caching) it's just slower on repeated use).

Because the Index version defines monotonic_increasing, monotonic_decreasing, and is_unique in a single call via _engine.

File: pandas/core/indexes/datetimes.py [hunk 19]
```diff
@@ -254,13 +257,11 @@ def __new__(cls, data=None,
             result.name = name
             return result
 
-        if not isinstance(data, (np.ndarray, Index, ABCSeries,
-                                 DatetimeArrayMixin)):
-            if is_scalar(data):
-                raise ValueError('DatetimeIndex() must be called with a '
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: My preference would have been to keep it a ValueError and change to a TypeError in a separate PR, but here we are... will add a note in Breaking Changes.

File: pandas/core/arrays/datetimes.py [hunk 20]
```diff
@@ -193,11 +196,16 @@ def _simple_new(cls, values, freq=None, tz=None, **kwargs):
...
+        if isinstance(values, (list, tuple)) or is_object_dtype(values):
+            values = cls._from_sequence(values, copy=copy)
+            # TODO: Can we set copy=False here to avoid re-coping?
+
         if tz is None and hasattr(values, 'tz'):
-            # e.g. DatetimeIndex
+            # e.g. DatetimeArray, DatetimeIndex
             tz = values.tz
 
+        # TODO: what about if freq == 'infer'?
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: > then we should also get the freq from the values as a "cheap" inference? 

That's what I'm thinking, yah

File: pandas/core/arrays/datetimes.py [hunk 21]
```diff
@@ -207,26 +215,46 @@ def __new__(cls, values, freq=None, tz=None, dtype=None):
         # if dtype has an embedded tz, capture it
         tz = dtl.validate_tz_from_dtype(dtype, tz)
 
-        if isinstance(values, DatetimeArrayMixin):
+        if lib.is_scalar(values):
+            raise TypeError(dtl.scalar_data_error(values, cls))
+        elif isinstance(values, ABCSeries):
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: I'll see if there is a graceful way to do this in the next pass (if I ever manage to catch up with all these comments!)

File: pandas/core/indexes/timedeltas.py (hunk 11 above)
Comment (jbrockmendel) [reply pandas/core/indexes/timedeltas.py]: I'm on board with the "leave for later" part of this

File: pandas/tests/arrays/test_datetimes.py (hunk 12 above)
Comment (jbrockmendel) [reply pandas/tests/arrays/test_datetimes.py]: > What is the reason you added handling of this to that method?

I'm not sure I understand the question.  Is there a reason not to?

File: pandas/tests/arrays/test_datetimes.py (hunk 13 above)
Comment (jbrockmendel) [reply pandas/tests/arrays/test_datetimes.py]: Hopefully I've answered this enough times in this thread.  I see no reason _not_ to handle object dtype in the DatetimeArray constructor.  Series and Index and DateFrame all handle object-dtype and lists; I find it counter-intuitive that we would have a small subset of classes that don't.

But in the short-run, we need to handle these cases if we want to share the extant tests (without significant overhaul) (which we do)

File: pandas/core/arrays/datetimes.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/core/arrays/datetimes.py]: > I don't share this opinion, 

Then please raise this in the appropriate issue, as we have been discussing this before (I think it is https://github.com/pandas-dev/pandas/issues/23212, although there is probably some more scattered discussion on other related PRs)

> would prefer to delay this discussion until it is absolutely necessary.

It is here that you are redesigning the constructors for the array refactor, IIUC, so if there is a time we should discuss it, it is now I think?

> Not if we want to share the extant arithmetic tests (which we do)

Can you clarify this a little bit? At what point do the arithmetic tests need to deal with array of objects? 
Eg for boxing the constructed values into Series/Index/Array, there a properly dtyped array can be used?

File: pandas/tests/arrays/test_datetimes.py (hunk 13 above)
Comment (jorisvandenbossche) [reply pandas/tests/arrays/test_datetimes.py]: See https://github.com/pandas-dev/pandas/pull/23493#discussion_r230850779 for my answer to a previous related comment

File: pandas/tests/arrays/test_datetimes.py (hunk 12 above)
Comment (jorisvandenbossche) [reply pandas/tests/arrays/test_datetimes.py]: Is there ever a case where call `_from_sequence` internally with 2D data? And if so, what case? 

I would think that at the point we internally call `_from_sequence`, we are sure it is a 1D array (eg as the result of some operation).

File: pandas/tests/arrays/test_datetimes.py (hunk 13 above)
Comment (jbrockmendel) [reply pandas/tests/arrays/test_datetimes.py]: Elsewhere in this thread you've said you're fine with bikeshedding this after we have a working implementation.  Has this changed in the last few minutes?

File: pandas/core/arrays/datetimes.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: > Can you clarify this a little bit? At what point do the arithmetic tests need to deal with array of objects?

The pertinent word here is "extant".  Many of the tests in tests/arithmetic pass a list into `tm.box_expected` or `klass`.

File: pandas/tests/arrays/test_datetimes.py (hunk 12 above)
Comment (jbrockmendel) [reply pandas/tests/arrays/test_datetimes.py]: You're far more competent than I am and can be counted on not to make this particular mistake; I cannot.

If we're not allowing object-dtype/lists in `__init__/__new__` then we are basically forcing users to use the (private!) `_from_sequence`; validation need to happen somewhere.

Comment (jorisvandenbossche) [reply pandas/tests/arrays/test_datetimes.py]: The `_from_sequence` method has a specific purpose in the EA interface: converting an array-like of scalars back to a proper Array (and should never be called by a user). Eg this method is used when doing an `astype(EAdtype)`, but I was thinking that it might well be that in those cases that we call it, it already has been checked the data is 1D. 

So I am simply honestly wondering if there is a specific case that you encountered now where this check was needed. 

But I suppose the answer is that you are using `_from_sequence` to validate generic user input, and there the user *can* pass 2D data and we should invalidate it.

However, it is not only you that do it, we actually already do it when doing `Series(..., dtype=EAdtype)` (where also user input is processed in `_from_sequence`), so indeed, we should probably have this check in general. 

So long story short: it indeed might make sense to add this type check. But then to repeat my initial comment: this should then be a test in the base EA tests to ensure all EAs do this properly?

File: pandas/core/arrays/datetimes.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/arrays/datetimes.py]: > _convert_listlike_datetimes calls ensure_object.

That's after an

```python
    # these are shortcutable
    if is_datetime64tz_dtype(arg):
        if not isinstance(arg, DatetimeIndex):
            return DatetimeIndex(arg, tz=tz, name=name)
        if tz == 'utc':
            arg = arg.tz_convert(None).tz_localize(tz)
        return arg

    elif is_datetime64_ns_dtype(arg):
        if box and not isinstance(arg, DatetimeIndex):
            try:
                return DatetimeIndex(arg, tz=tz, name=name)
            except ValueError:
                pass

        return arg
```

So those both avoid conversion to object.

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/arrays/datetimes.py]: `ExtensionArray._from_sequence` is for any sequence of scalar objects, including a ndarray with a specialized type like `datetime64[ns]`. It'll be used, for example, in `factorize(ExtensionArray)`.

File: pandas/core/arrays/datetimes.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/arrays/datetimes.py]: Ignoring the tests for a moment, I thought we were all on board with the goal of the `DatetimelikeArray.__init__` being no inference and no copy.

Back to the tests, it *looks* like you can you add an entry to `box_expected` for `DatetimeArray` to return `expected = DatetimeArray._from_sequence(expected)`?

**Comment by jbrockmendel** [comment]
@jorisvandenbossche Your attention has been specifically requested in #23514, whereas I am finding it increasingly frustrating.  I propose we step back from this conversation so I can spend some time addressing the subset of comments on which there is consensus.

**Comment by TomAugspurger** [comment]
Maybe we can spend a bit of time building consensus on a direction forward? I'll try to build my own thoughts here on a proposal, as a response / concurrence to 
https://github.com/pandas-dev/pandas/issues/23185#issuecomment-435592816.

FWIW, everything on the TimedeltaArray / DatetimeArray is on my critical path, so I'm going to prioritize reviewing your PRs over everything else.

**Comment by jorisvandenbossche** [comment]
> [Tom] I thought we were all on board with the goal of the DatetimelikeArray.init being no inference and no copy.

@jbrockmendel you have to see my comments in the same light. I was assuming we had this discussion and agreed on the design of the constructors, so I was wondering why we couldn't already follow that decision in this PR instead of deferring to a follow-up. 
So this mis-assumption might have made things more frustrating than needed for *both* of us.

> Your attention has been specifically requested in #23514, whereas I am finding it increasingly frustrating.

Just as for Tom, this PR (and all issues related to the EA refactor) is high on my priority list, so I will still put time on this topic (which is more than writing comments here). 
So you can expect a lot more comments to come (not necessarily now, but when you had the time to update the PR). But please don't see this as a bad sign, but rather as an opportunity to move forward quickly (and as an indicator of the importance of this refactor). Tom's PRs on the SparseArray and PeriodArray also saw a huge amount of comments and discussion, but I think that significantly improved the PRs and moved us relatively quickly to a shared understanding. 
And to be clear, it is no problem that it takes some time to process those comments.

If you have specific feedback on how I can make my comments less frustrating apart from the above, I am honestly all ears.

---

But indeed, let's first try to build consensus on the fundamental design questions. I would propose to not do that here on this PR but on the general issue https://github.com/pandas-dev/pandas/issues/23185 and the constructor split-off issue https://github.com/pandas-dev/pandas/issues/23212, so we can keep the discussions in this PR on technical implementation details of things we in general already agree on.

I would personally propose to have a chat about this, as a more effective way to discuss things. But @jbrockmendel, I think that only makes sense if you can fix the audio issues we were having last time. If that is not possible right now, I would propose to do it on chat at a given time to at least have it a bit more interactive than on github.

**Comment by jbrockmendel** [comment]
> If you have specific feedback on how I can make my comments less frustrating apart from the above, I am honestly all ears.

I appreciate the consideration.  At this point I think the short-term solution is for me to take a half-step back and disengage.  The request that you do the same was primarily so you don't feel that I'm becoming non-responsive.

Medium-run, I very much think the priority should be getting enough of this in place that we can get the tests working.  Later if you want to implement `datetime_array` mirroring `period_array` or rename `__init__` to `_from_sequence` (I'll be renaming `_from_sequence` to `_from_objects` or something) then that will be _entirely trivial_ to do.

I'll be going AFK (at least on this thread) for a few hours.

**Comment by jorisvandenbossche** [comment]
> So there will be at least one fixing-things PR before this comes back to the forefront.

I know I am not at all in the position to force things, so I will just give my opinion (and take it as that :-)): I think we should rather focus our effort on moving forward this PR (eg start trying to reach consensus on the discussion points above). 
It's really good you found those bugs, and we should certainly fix them, but those are not *critical* for the refactor, while the things discussed here are.

> In the interim, we can make incremental progress on orthogonal DTA/TDA code in #23415 and #23502.

I thought you mentioned #23415 was kind of dormant for now? On #23502 I added some additional comments.

**Comment by jbrockmendel** [comment]
> I thought you mentioned #23415 was kind of dormant for now?

You thought correctly, but it may be worthwhile to un-dormant-ize it so as to maintain forward momentum while bugs get sorted out.

> but those are not _critical_ for the refactor, while the things discussed here are.

Disagree on both counts.  Without the bugs being fixed, we can't institute meaningful tests for the array classes.  On the other hand, whether we rename `__init__/_simple_new` to `_from_sequence/__init__` can absolutely be postponed until after we have a fully-working implementation.

**Comment by jbrockmendel** [comment]
Closing.  I’ll salvage any useful tests in the PR that comes after #23539.
